From 7662587f10552d4aad04ba2b95e5643ae2107d6f Mon Sep 17 00:00:00 2001
From: Tom Deseyn <tom.deseyn@gmail.com>
Date: Wed, 19 Oct 2022 15:21:40 +0200
Subject: [PATCH 1/1] Enabled source-building with mono runtime on any
 architecture.

Adds an argument to the top-level build script to use the mono
runtime, and sets the flags needed for the different repos
to work with mono.
---
 eng/SourceBuild.props                         |   1 +
 .../tarball/content/Directory.Build.props     |   5 +
 src/SourceBuild/tarball/content/build.sh      |   4 +
 .../content/repos/Directory.Build.props       |   1 +
 ...d-support-building-with-mono-runtime.patch |  24 ++
 ...d-support-building-with-mono-runtime.patch |  24 ++
 ...d-support-building-with-mono-runtime.patch | 324 ++++++++++++++++++
 7 files changed, 383 insertions(+)
 create mode 100644 src/SourceBuild/tarball/patches/aspnetcore/0002-source-build-support-building-with-mono-runtime.patch
 create mode 100644 src/SourceBuild/tarball/patches/runtime/0001-source-build-support-building-with-mono-runtime.patch
 create mode 100644 src/SourceBuild/tarball/patches/sdk/0001-source-build-support-building-with-mono-runtime.patch

diff --git a/Directory.Build.props b/Directory.Build.props
index 860785575..982fb7f07 100644
--- a/Directory.Build.props
+++ b/Directory.Build.props
@@ -23,6 +23,11 @@
     <Platform Condition="'$(Platform)' == ''">x64</Platform>
 
     <UseStableVersions Condition="'$(UseStableVersions)' == ''">false</UseStableVersions>
+
+    <SourceBuildUseMonoRuntime Condition="'$(SourceBuildUseMonoRuntime)' == ''">false</SourceBuildUseMonoRuntime>
+    <!-- These architectures are only supported with mono runtime -->
+    <SourceBuildUseMonoRuntime Condition="'$(BuildArchitecture)' == 's390x'">true</SourceBuildUseMonoRuntime>
+    <SourceBuildUseMonoRuntime Condition="'$(BuildArchitecture)' == 'ppc64le'">true</SourceBuildUseMonoRuntime>
   </PropertyGroup>
 
   <!-- This repo's projects are entirely infrastructure and do not ship. -->
diff --git a/repos/Directory.Build.props b/repos/Directory.Build.props
index 860785575..982fb7f07 100644
--- a/repos/Directory.Build.props
+++ b/repos/Directory.Build.props
@@ -143,6 +143,7 @@
     <StandardSourceBuildArgs>$(StandardSourceBuildArgs) /p:AdditionalSourceBuiltNupkgCacheDir="$(SourceBuiltPackagesPath)"</StandardSourceBuildArgs>
     <StandardSourceBuildArgs>$(StandardSourceBuildArgs) /p:ReferencePackageNupkgCacheDir="$(ReferencePackagesDir)"</StandardSourceBuildArgs>
     <StandardSourceBuildArgs>$(StandardSourceBuildArgs) /p:PreviouslySourceBuiltNupkgCacheDir="$(PrebuiltSourceBuiltPackagesPath)"</StandardSourceBuildArgs>
+    <StandardSourceBuildArgs>$(StandardSourceBuildArgs) /p:SourceBuildUseMonoRuntime=$(SourceBuildUseMonoRuntime)</StandardSourceBuildArgs>
 
     <StandardSourceBuildCommand>$(ProjectDirectory)\build$(ShellExtension)</StandardSourceBuildCommand>
   </PropertyGroup>

diff --git a/build.sh b/build.sh
index e899d2f80..17699c78c 100755
--- a/build.sh
+++ b/build.sh
@@ -11,6 +11,7 @@ usage() {
     echo "  --run-smoke-test                   don't build; run smoke tests"
     echo "  --with-packages <dir>              use the specified directory of previously-built packages"
     echo "  --with-sdk <dir>                   use the SDK in the specified directory for bootstrapping"
+    echo "  --use-mono-runtime                 output uses the mono runtime"
     echo "use -- to send the remaining arguments to MSBuild"
     echo ""
 }
@@ -65,6 +66,9 @@ while :; do
             fi
             shift
             ;;
+        --use-mono-runtime)
+            MSBUILD_ARGUMENTS+=( "/p:SourceBuildUseMonoRuntime=true" )
+            ;;
         --)
             shift
             echo "Detected '--': passing remaining parameters '$@' as build.sh arguments."
-- 
2.36.3

